### PR TITLE
Canonicalize subtraction with constant on the right + FinalOptimize

### PR DESCRIPTION
**Pull request by MaxGraey**
canonicalize stage:
`x - C` => `x + (-C)`

this PR also exposed finalize stage which cancel previous transform later. "signed LEB optimizations" was moved into this final stage as well. In future we could use more special rules for this finilized stage like 
`(signed)x < 0` -> `x >>> 31`, 
`bool(x != -1)` -> `bool(x ^ -1)` 
and etc.

Suggested here: #3320

### CODE REVIEW

**Comment by MaxGraey** [comment]
@tlively @kripken could you take a look at this?

File: src/passes/OptimizeInstructions.cpp [hunk 1]
```diff
@@ -151,12 +151,16 @@ struct OptimizeInstructions
   : public WalkerPass<
       PostWalker<OptimizeInstructions,
                  UnifiedExpressionVisitor<OptimizeInstructions>>> {
-  bool isFunctionParallel() override { return true; }
-
```
Comment (kripken) [inline src/passes/OptimizeInstructions.cpp:155]: please don't change the order of these, the previous order was consistent across the codebase.

**Review by kripken — COMMENTED** [review]
I think these two ideas are separate, aren't they? A separate PR for the canonicalization, and one for the finalization, would be good.

The canonicalization seems obviously useful, while the finalization I am still not sure what the idea is. So separating the discussion would be good.

**Comment by MaxGraey** [comment]
> I think these two ideas are separate, aren't they? A separate PR for the canonicalization, and one for the finalization, would be good.

Finalization strictly necessary otherwise `optimization with signed LEB` should be disabled due to it conflicted with this canonization

**Comment by MaxGraey** [comment]
Fuzzed:
```
Invocations so far:
   FuzzExec: 8168
   CompareVMs: 1441
   CheckDeterminism: 714
   Wasm2JS: 856
   Asyncify: 1357

ITERATION: 9024
```

File: src/passes/OptimizeInstructions.cpp (hunk 1 above)
Comment (MaxGraey) [reply src/passes/OptimizeInstructions.cpp:155]: Got it. Reverted!

**Comment by MaxGraey** [comment]
Btw here https://github.com/WebAssembly/binaryen/compare/master...MaxGraey:1f508ce6d198d402974ec8e125ac4f6365cf6167 you could look at this PR without finalize stage (only canonicalization)

**Comment by kripken** [comment]
I see, thanks @MaxGraey . Yes, the two things are connected, good point.

I think we can maybe do something less complicated than adding a finalize variation of the entire pass, though. That has some downsides, like people needing to remember to call both.

How about after the pass is all done, it runs another final operation to do final fixups. It could be like [`FinalOptimizer` is used in RemoveUnusedBrs](https://github.com/WebAssembly/binaryen/blob/040a7cc3721292669609cc8020bd11ad6930d321/src/passes/RemoveUnusedBrs.cpp#L734).

File: src/passes/OptimizeInstructions.cpp [hunk 2]
```diff
@@ -912,7 +920,13 @@ struct OptimizeInstructions
     if (binary->left->is<Const>() && !binary->right->is<Const>()) {
       return swap();
     }
-    if (binary->right->is<Const>()) {
+    if (auto* c = binary->right->dynCast<Const>()) {
+      // x - C  ==>   x + (-C)
+      // Prefer use addision operation if constant on the right.
```
Comment (kripken) [inline src/passes/OptimizeInstructions.cpp:925]:
```suggestion
      // Prefer use addition if there is a constant on the right.
```

File: src/passes/OptimizeInstructions.cpp [hunk 3]
```diff
@@ -1208,12 +1262,6 @@ struct OptimizeInstructions
             replaceCurrent(curr->left);
             return;
           }
-        } else if (curr->op == Abstract::getBinary(type, Abstract::Sub)) {
-          // we must leave a left zero, as it is how we negate ints
-          if (right && right->value.isZero()) {
-            replaceCurrent(curr->left);
-            return;
-          }
```
Comment (kripken) [inline src/passes/OptimizeInstructions.cpp:1216]: why was this removed?

File: src/passes/OptimizeInstructions.cpp [hunk 4]
```diff
@@ -2390,6 +2398,10 @@ struct OptimizeInstructions
   }
 
   bool isSymmetricOrRelational(Binary* binary) {
+    if ((binary->op == SubInt32 || binary->op == SubInt64) &&
+        binary->right->is<Const>() && !binary->left->is<Const>()) {
+      return true;
```
Comment (kripken) [inline src/passes/OptimizeInstructions.cpp:2403]: this change looks confusing: `Sub` is neither symmetric nor relational. Perhaps we should rename the function, if it makes sense to consider a `Sub` as well in all the callers? (Why does that make sense?)

Comment (MaxGraey) [reply src/passes/OptimizeInstructions.cpp:2403]: It's required for pass this guard for `x - C` case which will be canonicalized to `x + (-C)`. So perhaps rename this function to something like `bool shouldCanonicalize(Binary* binary)` wdyt?

File: src/passes/OptimizeInstructions.cpp (hunk 3 above)
Comment (MaxGraey) [reply src/passes/OptimizeInstructions.cpp:1216]: it's checked `x - 0` which with this canonicalization always should be `x + 0` handled in additional arm

**Comment by MaxGraey** [comment]
> How about after the pass is all done, it runs another final operation to do final fixups. It could be like FinalOptimizer is used in RemoveUnusedBrs.

Yeah, this makes sense!

**Comment by MaxGraey** [comment]
@kripken I refactored with FinalOptimizer. Could you take a look?

File: src/passes/OptimizeInstructions.cpp (hunk 4 above)
Comment (kripken) [reply src/passes/OptimizeInstructions.cpp:2424]: Yes, that name sounds better!

**Comment by kripken** [comment]
> Yeah, this makes sense! But as default pipeline called optimize-institutions several times and in case with built-in Post Opt Pass it will cause to unnecessary work a little. Like:

That's true, and it's a downside in RemoveUnusedBrs as well. But it is simpler, and the extra work is very fast, so I think it's worth doing that way. But, new ideas for destructive things like you mention may be a reason to reconsider it later.

File: src/passes/OptimizeInstructions.cpp [hunk 5]
```diff
@@ -133,6 +133,69 @@ struct LocalScanner : PostWalker<LocalScanner> {
   }
 };
 
+// perform some final optimizations
```
Comment (kripken) [inline src/passes/OptimizeInstructions.cpp:136]: Please add
```cpp
namespace {
```
before, and
```cpp
} // anonymous namespace
```
after this class. Without them, if another file defines something with the same name, they will be linked together. (This is only a problem in the top level, that is, the global scope.)

File: src/passes/OptimizeInstructions.cpp [hunk 6]
```diff
@@ -133,6 +133,69 @@ struct LocalScanner : PostWalker<LocalScanner> {
...
+struct FinalOptimizer
+  : public WalkerPass<
+      PostWalker<FinalOptimizer, UnifiedExpressionVisitor<FinalOptimizer>>> {
```
Comment (kripken) [inline src/passes/OptimizeInstructions.cpp:139]: this does not need to be a Pass. It just needs to do a walk, so it can be like in RemoveUnusedBrs,
```cpp
struct FinalOptimizer : public PostWalker<FinalOptimizer> {
```

File: src/passes/OptimizeInstructions.cpp [hunk 7]
```diff
@@ -133,6 +133,69 @@ struct LocalScanner : PostWalker<LocalScanner> {
...
+  FinalOptimizer(const PassOptions& passOptions) : passOptions(passOptions) {}
+
+  void visitExpression(Expression* curr) {
+    // if this contains dead code, don't bother trying to optimize it, the type
+    // might change (if might not be unreachable if just one arm is, for
+    // example). this optimization pass focuses on actually executing code. the
+    // only exceptions are control flow changes
+    if (curr->is<Const>() || curr->is<Call>() || curr->is<Nop>() ||
+        (curr->type == Type::unreachable && !curr->is<Break>() &&
+         !curr->is<Switch>() && !curr->is<If>())) {
```
Comment (kripken) [inline src/passes/OptimizeInstructions.cpp:151]: I think all these checks are not needed. They can be removed, and this function merged with `optimize`, so there is just one function, `visitExpression`.

File: src/passes/OptimizeInstructions.cpp (hunk 3 above)
Comment (kripken) [reply src/passes/OptimizeInstructions.cpp:1216]: Good point, I think you're right. But please add a comment in the code for add,
```
Note that we don't need to handle subtraction: canonicalization avoids a zero on the right anyhow, and a zero on the left can't be removed (it is how we negate ints).
```
Without the comment, it might be confusing why subtraction is not handled.

File: src/passes/OptimizeInstructions.cpp [hunk 8]
```diff
@@ -153,10 +216,13 @@ struct OptimizeInstructions
                  UnifiedExpressionVisitor<OptimizeInstructions>>> {
   bool isFunctionParallel() override { return true; }
 
-  Pass* create() override { return new OptimizeInstructions; }
+  Pass* create() override { return new OptimizeInstructions(finalize); }
 
+  bool finalize = false;
```
Comment (kripken) [inline src/passes/OptimizeInstructions.cpp:221]: as mentioned, I think we do not need the `finalize` stuff. We can consider adding it later if speed or future optimizations are an issue, I think?

**Comment by MaxGraey** [comment]
So with recent changes we have following approach:
```cpp
add("optimize-instructions");
...
add("optimize-instructions");
...
add("optimize-instructions-finalize"); // last one
```

before was:
```cpp
add("optimize-instructions");
...
add("optimize-instructions");
...
add("optimize-instructions"); // last one
add("optimize-instructions-finalize"); // extra final optimization only
```

File: src/passes/OptimizeInstructions.cpp [hunk 9]
```diff
@@ -133,6 +133,69 @@ struct LocalScanner : PostWalker<LocalScanner> {
...
 
+// perform some final optimizations
+struct FinalOptimizer
+  : public WalkerPass<
+      PostWalker<FinalOptimizer, UnifiedExpressionVisitor<FinalOptimizer>>> {
+  const PassOptions& passOptions;
+
+  FinalOptimizer(const PassOptions& passOptions) : passOptions(passOptions) {}
+
+  void visitExpression(Expression* curr) {
```
Comment (kripken) [inline src/passes/OptimizeInstructions.cpp:144]: Oh, and this can be `visitBinary(Binary* curr)` as that's the only thing we care about.

**Comment by kripken** [comment]
@MaxGraey I agree the recent changes make it simpler, but I'm still not convinced we need the finalize stuff, for the reasons I've said (adds a little complexity, saves only a tiny amount of work).

**Comment by MaxGraey** [comment]
Alright, I'll remove `optimize-instructions-finalize`

File: src/passes/OptimizeInstructions.cpp (hunk 6 above)
Comment (MaxGraey) [reply src/passes/OptimizeInstructions.cpp:139]: I tried this but unfortunately this not works for me (visitExpression just not called) so I switched to `WalkerPass<PostWalker<>>`

File: src/passes/OptimizeInstructions.cpp (hunk 3 above)
Comment (MaxGraey) [reply src/passes/OptimizeInstructions.cpp:1216]: Done

File: src/passes/OptimizeInstructions.cpp (hunk 5 above)
Comment (MaxGraey) [reply src/passes/OptimizeInstructions.cpp:137]: Done!

File: src/passes/OptimizeInstructions.cpp (hunk 9 above)
Comment (MaxGraey) [reply src/passes/OptimizeInstructions.cpp]: Done!

File: src/passes/OptimizeInstructions.cpp (hunk 6 above)
Comment (MaxGraey) [reply src/passes/OptimizeInstructions.cpp:140]: Hmm, but now with `visitBinary` it works!

**Comment by MaxGraey** [comment]
I finished with refactoring. @kripken could you take a look please?

File: src/passes/OptimizeInstructions.cpp [hunk 10]
```diff
@@ -133,6 +133,57 @@ struct LocalScanner : PostWalker<LocalScanner> {
...
 
+namespace {
+// perform some final optimizations
+struct FinalOptimizer : public PostWalker<FinalOptimizer> {
+  const PassOptions& passOptions;
+
+  FinalOptimizer(const PassOptions& passOptions) : passOptions(passOptions) {}
+
+  void visitBinary(Binary* curr) {
+    if ((curr = optimize(curr))) {
```
Comment (kripken) [inline src/passes/OptimizeInstructions.cpp:144]:
```suggestion
    if (auto* replacement =optimize(curr)) {
```

File: src/passes/OptimizeInstructions.cpp [hunk 11]
```diff
@@ -133,6 +133,57 @@ struct LocalScanner : PostWalker<LocalScanner> {
...
+      replaceCurrent(curr);
```
Comment (kripken) [inline src/passes/OptimizeInstructions.cpp:145]:
```suggestion
      replaceCurrent(replacement);
```

File: src/passes/OptimizeInstructions.cpp [hunk 12]
```diff
@@ -167,6 +218,12 @@ struct OptimizeInstructions
     }
     // main walk
     super::doWalkFunction(func);
+    // do final optimizations if finalize is set
```
Comment (kripken) [inline src/passes/OptimizeInstructions.cpp:221]: can delete this comment
```suggestion
```

File: src/passes/OptimizeInstructions.cpp [hunk 13]
```diff
@@ -167,6 +218,12 @@ struct OptimizeInstructions
...
+    {
+      FinalOptimizer optimizer(getPassOptions());
+      optimizer.setModule(getModule());
```
Comment (kripken) [inline src/passes/OptimizeInstructions.cpp:224]: the module is not used there, so this can be deleted
```suggestion
```

File: src/passes/OptimizeInstructions.cpp [hunk 14]
```diff
@@ -1199,6 +1259,9 @@ struct OptimizeInstructions
         auto type = curr->type;
         auto* left = curr->left->dynCast<Const>();
         auto* right = curr->right->dynCast<Const>();
+        // Note that we don't need to handle subtraction:
+        // canonicalization avoids a zero on the right anyhow, and a
+        // zero on the left can't be removed (it is how we negate ints).
```
Comment (kripken) [inline src/passes/OptimizeInstructions.cpp:1264]: reading this now, I'm confused. why does canonicalization avoid a zero on the right? let's improve the comment, because if I can't figure it out now after just thinking about this a day or two ago, it will be confusing for other people I think...

I think a better comment could be `Canonicalization prefers an add instead of a subtract wherever possible. That prevents a subtracted constant on the right, as it would be added. And for a zero on the left, it can't be removed (it is how we negate ints).`

File: src/wasm2js.h [hunk 15]
```diff
@@ -338,6 +338,8 @@ Ref Wasm2JSBuilder::processWasm(Module* wasm, Name funcName) {
       }
       runner.addDefaultOptimizationPasses();
       runner.add("avoid-reinterprets");
+    } else {
+      runner.add("optimize-instructions");
```
Comment (kripken) [inline src/wasm2js.h:342]: why is this added? if it makes sense, let's do it in a separate PR with a full explanation.

Comment (MaxGraey) [reply src/wasm2js.h:342]: Sure. Will remove this

**Review by kripken — APPROVED** [review]
